cpp/oneapi/dal/backend/primitives/reduction/functors.hpp

```diff
@@ -51,6 +51,38 @@ struct square {
     }
 };
 
+template <typename T>
+struct isinfornan {
+    using tag_t = reduce_unary_op_tag;
+    bool operator()(const T& arg) const {
+#ifdef ONEDAL_DATA_PARALLEL
+        return std::numeric_limits<T>::infinity ==
+               sycl::ext::oneapi::bit_and<T>(std::numeric_limits<T>::infinity, arg);
+#else
+        return std::numeric_limits<T>::infinity ==
+               std::bit_and(std::numeric_limits<T>::infinity, arg);
+#endif
+    }
+};
+
+template <typename T>
+struct isinf {
+    using tag_t = reduce_unary_op_tag;
+    // the bitwise not of negative infinity is == mantissa with all ones
+    constexpr static inline T fracmask = ~(-std::numeric_limits<T>::infinity);
+    bool operator()(const T& arg) const {
+#ifdef ONEDAL_DATA_PARALLEL
+        return std::numeric_limits<T>::infinity ==
+                   sycl::ext::oneapi::bit_and<T>(std::numeric_limits<T>::infinity, arg) &&
+               0 == sycl::ext::oneapi::bit_and<T>(fracmask, arg);
+#else
+        return std::numeric_limits<T>::infinity ==
+                   std::bit_and(std::numeric_limits<T>::infinity, arg) &&
+               0 == std::bit_and(fracmask, arg);
```

Comment:
I think this can be simplified to just check for positive and negative infinities:
```suggestion
        return (std::numeric_limits<T>::infinity == arg) ||  (-std::numeric_limits<T>::infinity == arg);
```